File: packages/google_mlkit_face_mesh_detection/ios/Classes/GoogleMlKitFaceMeshDetectionPlugin.m
#import "GoogleMlKitFaceMeshDetectionPlugin.h"
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_face_mesh_detector"
#define startFaceMeshDetector @"vision#startFaceMeshDetector"
#define closeFaceMeshDetector @"vision#closeFaceMeshDetector"

@implementation GoogleMlKitFaceMeshDetectionPlugin {
    NSMutableDictionary *instances;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitFaceMeshDetectionPlugin* instance = [[GoogleMlKitFaceMeshDetectionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startFaceMeshDetector]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeFaceMeshDetector]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    // TODO: waiting for Google to release Face Mesh api for iOS
    // https://developers.google.com/ml-kit/vision/face-mesh-detection
    result(FlutterMethodNotImplemented);
}

@end

File: packages/google_mlkit_face_mesh_detection/ios/Classes/GoogleMlKitFaceMeshDetectionPlugin.swift
import Flutter
import google_mlkit_commons

@objc
public class GoogleMlKitFaceMeshDetectionPlugin: NSObject, FlutterPlugin {
  private var instances: [String: Any] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_face_mesh_detector",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitFaceMeshDetectionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startFaceMeshDetector":
      handleDetection(call: call, result: result)
    case "vision#closeFaceMeshDetector":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    // swiftlint:disable:next todo
    // TODO: waiting for Google to release Face Mesh API for iOS
    result(FlutterMethodNotImplemented)
  }
}
